[WIP] sys/linux, tools: add ksmbd (SMB3 server) stateful fuzzing

yskzalloc opened the pull request:
Add ksmbd, the in-kernel SMB3 server, as a syzkaller target: the SMB2 protocol grammar, the transport syscalls (enabled in vminfo), and the image setup that brings the server up so the descriptions reach a live, configured daemon.
    
### Transport (sys/linux/ksmbd.txt):
    
`socket$ksmbd -> connect$ksmbd -> sendto$ksmbd`
    
 A sock_ksmbd resource is threaded through connect and sendto, so a  program that issues several sendto$ksmbd() on the same socket forms a stateful SMB session: ksmbd keeps per-connection state (negotiated dialect, authenticated session, opened FIDs, granted oplocks/leases) across the PDUs. The file documents this model, and its one gap: ksmbd assigns SessionId/TreeId/Persistent-VolatileFileId in the response PDUs, which syzkaller does not parse, so the server-assigned ids cannot yet be threaded into later PDUs (they are seeded with the SMB2 all-ones sentinel and 0 so brute/dataflow can still land valid ids). Closing that gap needs a send+recv pseudo-syscall that returns the ids as resources.
    
### Protocol grammar:
    
- smb2_hdr with the ProtocolId/StructureSize constants and Command/Flags as flag sets;
- request bodies for negotiate, session-setup, tree-connect, create, close, flush, read, write, lock, ioctl, cancel, echo, query-directory, change-notify, query-info, set-info and oplock-break, dispatched by a union keyed on hdr.Command;
- structured create contexts (durable-handle DHnQ/DH2Q, lease RqLs, and generic tagged contexts) instead of an opaque blob;
- an FSCTL input union for IOCTL (copychunk, validate-negotiate-info);
- a set-info file-information union (rename, disposition, end-of-file, allocation, basic);
- flag sets for the access mask, share access, create disposition and options, oplock level, impersonation, file attributes, info classes, dialects and FSCTL codes, so directed mutation lands valid values.
    
### Image setup:
    
`create-image.sh --add-ksmbd` installs ksmbd-tools plus the RDMA (rdma-core, iproute2) and Kerberos (krb5-*) userspace, stages the self-contained `tools/ksmbd_setup.sh`, and enables it as a systemd.

tarasmadan commented:
Nice!
And we don't have smb2 subsystem defined to track the coverage progress. It is not a blocker, just a TODO.

tarasmadan commented:
We have some coverage from fs/smb/client [here](https://syzkaller.appspot.com/upstream/graph/coverage_heatmap?period=month&subsystem=cifs).
We don't have any coverage from fs/smb/server because SMB_SERVER(at least) config option isn't enabled.

a-nogikh commented on executor/common_linux.h:
```diff
@@ -5830,3 +5830,64 @@ static long syz_pidfd_open(volatile long pid, volatile long flags)
+
+#define KSMBD_BUF_SIZE 16000
+
+static long syz_ksmbd_send_req(volatile long a0, volatile long a1, volatile long a2, volatile long a3)
```
Do you plan to read out the ksmdb responses and somehow act upon them?

If not, we probably don't really need to introduce a pseudo system call at all. If you pass the kcov remote reference the standard way, all that remains are:
* A `socket` call.
* A `connect` call with the specific arguments.
* A `write` call.

These can be expressed in syzkaller descriptions and e.g. tied together by defining some custom `resource sock_ksmbd[sock]` resource.

dvyukov replied:
+1 for not adding a pseudo syscall. We just need to describe the tcp address, and add a special resource for the socket to tie all syscall together (socket/connect/sendmsg).

It has an additional advantage of giving more freedom to the fuzzer and exploring more scenarios. E.g. sending messages in parallel, calling shutdown, etc.

a-nogikh commented on executor/common_linux.h:
```diff
@@ -5830,3 +5830,64 @@ static long syz_pidfd_open(volatile long pid, volatile long flags)
+
+	if (a0 == 0 || a1 == 0) {
+		debug("[!]{syz_ksmbd_send_req} param empty\n");
+		return -7;
```
For the fuzzer, it's not really important to distinguish between these specific failure paths. I would just return -1 for all of them.

a-nogikh commented on pkg/vminfo/linux_syscalls.go:
```diff
@@ -103,6 +103,7 @@ var linuxSyscallChecks = map[string]func(*checkContext, *prog.Syscall) string{
 	"syz_socket_connect_nvme_tcp": linuxSyzSocketConnectNvmeTCPSupported,
 	"syz_pidfd_open":              alwaysSupported,
 	"syz_create_resource":         alwaysSupported,
+	"syz_ksmbd_send_req":          alwaysSupported,
```
I guess that for most users it will actually be not supported, so it would be nice to be more concrete here.

What is the simplest way of determining whether the server exists?

a-nogikh commented on sys/linux/ksmbd.txt:
```diff
@@ -0,0 +1,168 @@
+#	Id		_union_smb_req_id
+#	SessionId	int64
+#	Signature	array[int8, 16]
+#} [packed]
```
If these structs are not necessary, let's just drop them from the final descriptions.

a-nogikh commented on sys/linux/ksmbd.txt:
```diff
@@ -0,0 +1,168 @@
+
+smb2_req {
+	req_base	_union_smb2_req
+	req_andx1	optional[_union_smb2_req]
```
Nit: if that depends on the specific values set in the header, please note that there's a way to express that in syzlang as well:

https://github.com/google/syzkaller/blob/master/docs/syscall_descriptions_syntax.md#conditional-fields

yskelg commented:
Thank you all for your excellent comments. They have been helpful in guiding the work in a positive direction, and I will do my best to incorporate the points you mentioned.

yskzalloc committed:
sys/linux, tools: add ksmbd (SMB3 server) fuzzing target

Add ksmbd, the in-kernel SMB3 server, as a syzkaller target: the SMB2
protocol grammar, the transport syscalls (enabled in vminfo), and the
image setup that brings the server up so the descriptions reach a live,
configured daemon.

Transport (sys/linux/ksmbd.txt):

  socket$ksmbd -> connect$ksmbd -> sendto$ksmbd

A sock_ksmbd resource is threaded through connect and sendto, so a
program that issues several sendto$ksmbd() on the same socket forms a
stateful SMB session: ksmbd keeps per-connection state (negotiated
dialect, authenticated session, opened FIDs, granted oplocks/leases)
across the PDUs. The file documents this model, and its one gap: ksmbd
assigns SessionId/TreeId/Persistent-VolatileFileId in the response PDUs,
which syzkaller does not parse, so the server-assigned ids cannot yet be
threaded into later PDUs (they are seeded with the SMB2 all-ones sentinel
and 0 so brute/dataflow can still land valid ids). Closing that gap needs
a send+recv pseudo-syscall that returns the ids as resources.

Protocol grammar:

- smb2_hdr with the ProtocolId/StructureSize constants and Command /
  Flags as flag sets;
- request bodies for negotiate, session-setup, tree-connect, create,
  close, flush, read, write, lock, ioctl, cancel, echo,
  query-directory, change-notify, query-info, set-info and
  oplock-break, dispatched by a union keyed on hdr.Command;
- structured create contexts (durable-handle DHnQ/DH2Q, lease RqLs, and
  generic tagged contexts) instead of an opaque blob;
- an FSCTL input union for IOCTL (copychunk, validate-negotiate-info);
- a set-info file-information union (rename, disposition, end-of-file,
  allocation, basic);
- flag sets for the access mask, share access, create disposition and
  options, oplock level, impersonation, file attributes, info classes,
  dialects and FSCTL codes, so directed mutation lands valid values.

Image setup:

create-image.sh --add-ksmbd installs ksmbd-tools plus the RDMA
(rdma-core, iproute2) and Kerberos (krb5-*) userspace, stages the
self-contained tools/ksmbd_setup.sh, and enables it as a systemd
service so ksmbd comes up on every boot. mountd daemon then persists
independently of executor restarts. ksmbd_setup.sh writes the
server config, loads the module, provisions the fuzz user, sets up
Software RDMA (SIW/RXE) for SMB Direct and a local KDC for the krb5
auth path, then starts ksmbd.mountd. The RDMA and KDC steps are
non-fatal, so a kernel without SMB Direct still gets a working TCP
server on 127.0.0.1:445.

Cc: linux-cifs@vger.kernel.org
Cc: notselwyn@pwning.tech
Cc: Namjae Jeon <linkinjeon@kernel.org>
Signed-off-by: Yunseong Kim <yunseong.kim@est.tech>

yskzalloc commented:
Thank you to everyone for the reviews and feedback.

My understanding is that fuzzing each ksmbd call flow individually would be the proper approach, so I tried to model the call flows at a finer granularity, using "grains" as a metaphor and LLVM LibFuzzer as a mutation engine: https://github.com/yskzalloc/kcov-dataflow/blob/main/ksmbd/GRAIN.md#11-per-grain-diagrams-by-command

This work is not as extensive as syzkaller, but it helped uncover more than 10 ksmbd bugs.

I want to know the right way of applying this idea in syzkaller. There are simply too many call flows to describe as separate pseudo-syscalls, and maintaining them would be difficult. Is there any way to model or generate these call flows more efficiently in syzkaller without defining a pseudo-syscall for every single flow?
